docker_common.py: Support specifying docker images by content hash only.

wrouesnel opened the pull request:
##### SUMMARY
When specifying a docker image to any of the modules sharing the `docker_common.py` family of functions, passing in an image hash like `sha256:f49eec89601e8484026a8ed97be00f14db75339925fad17b440976cffcbfb88a` was wrongly interpreted as specifying an upstream repository on dockerhub.

This patch modifies the `find_image` and `_image_get` functions to correctly detect reserved names like this (of which` sha256` is currently the only one) and only search locally for the image.

This allows predictable use of docker when an image may be built and uploaded locally without a repository - i.e. in the case when a docker export tar is imported.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
docker_common

##### ANSIBLE VERSION
```
ansible 2.4.0 (support_sha265_ids 320f5fadb4) last updated 2017/06/16 04:06:47 (GMT +1100)
  config file = None
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible-work/ansible/lib/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.12+ (default, Sep 17 2016, 12:08:02) [GCC 6.2.0 20160914]

```


##### ADDITIONAL INFORMATION

wrouesnel committed:
docker_common.py: Support specifying docker images by content hash only.

This patch adds support for detecting the reserved "sha256" name to
the common image search functions for docker. This enables launching containers
identified by a local image content hash, as is the case when uploading
a tar file containing an untagged image.

wrouesnel commented:
@kassiansun @chouseknecht Pinging you guys for a review.

kassiansun commented on lib/ansible/module_utils/docker_common.py:
```diff
@@ -410,12 +411,20 @@ def _image_lookup(self, name, tag):
+        lookup = "%s:%s" % (name, tag)
+        if tag and name in DOCKER_RESERVED_IMAGE_NAMES:
+            for image in response:
+                if image['Id'] == lookup:
```
Is `image[Id]` a long `sha256` hash like `d3d92961e7c244734d6bb307dfecab651e96d4b6e18f4390f734cca477453302`? 
Should we allow users to specify it with a short one like `git`?.

wrouesnel replied:
Hm interesting thought. Docker command line does allow short form IDs - ala on my machine `docker inspect 8fc` works because it's unique to one image currently on my system.

This currently does not work with the code in this module (complains "registry returned more then 1 result") and I'm inclined to say that for an ansible module you should specify an exact hash if you're going this path, but the behaviour should definitely be defined.

kassiansun replied:
`docker_py` use `id[:10]` as `short_id`, but `docker images` shows a 12-char version. I think there's no "standard" short version of `image[Id]`.

"Registry returned more than one result for sha256:8fc" should be good enough. Users know what they're doing.

wrouesnel replied:
Fixed (2 issues - images needed to be blanked, and we check for all matches in images which allows partials).

I feel like a `[WARNING]` should pop up when you do that, but there's no existing plumbing in the AnsibleDockerClient to talk to the calling module logger, so it would be a bigger change (possibly for a separate PR).

kassiansun replied:
> I feel like a [WARNING] should pop up when you do that

When specify a hash string that matches more than one image? We should `fail` immediately, this is users' fault.

wrouesnel replied:
Already fixed then (it does fail if you match more then 1).

chouseknecht commented on lib/ansible/module_utils/docker_common.py:
```diff
@@ -381,7 +382,7 @@ def find_image(self, name, tag):
         images = self._image_lookup(name, tag)
-        if len(images) == 0:
+        if len(images) == 0 and name not in DOCKER_RESERVED_IMAGE_NAMES:
             # In API <= 1.20 seeing 'docker.io/<name>' as the name of images pulled from docker hub
```
It took me a while to figure how 'name' could be in DOCKER_RESERVED_IMAGE_NAMES. And then it hit me, `docker_container`, and probably `docker_image`, split the `image` parameter on ':', and assume that the first element is 'name', and the second element is a 'tag'. 

I think it would be preferable to have the code recognize when it receives an ID or partial ID value, and search accordingly. It would be nice if this could work with partial IDs, but this code will not support that.

wrouesnel replied:
Reasonable - I'll revisit this and chase down the places the name splitting is happening.

kassiansun replied:
`sha256` is reserved by docker, it's good enough to detect the usage of hash id.

chouseknecht replied:
It is. So rather than passing name and tag to this method, perhaps modules should pass the full name option (i.e., don't do the split), and this method should inspect the name, and decide how to best find the image.

wrouesnel replied:
The problem with unpicking the usage of `find_image` is that its used in 1 or 2 places in `docker_container` and *everywhere* in `docker_image_facts` and `docker_image` which makes this rather a large change, unless the interface just gets bifurcated for the time being between "things which deal with repository:tag and things which deal with a name"

chouseknecht reviewed (changes requested):
Let's not rely on the name being `sha256`. Let's instead recognize when we're being asked to identify an image based on an ID or partial ID.

chouseknecht commented:
@wrouesnel 
@kassiansun 

Would love to hear your opinions and ideas on the `docker_container` module at the Ansible Container Working Group. Would you be able to join us? We meet every Monday at 12:00 PM Eastern. Would really appreciate it if you could stop by and help us out. 

[View more info about the Ansible Container Working Group here](https://github.com/ansible/community/tree/master/group-container).

wrouesnel committed:
docker-common.py: Support finding images by RepoDigest.

The current image comparison support cannot deal with images specified by
RepoDigest and will fail to find a changed image (due to being unable to find
any image).

wrouesnel commented:
Updating this PR to deal with another issue related to content-hashes - the same image comparison function can't handle SHA256 RepoDigests and will fail to update/restart a container when the image is changed.

felixfontein commented:
This has been partially fixed in #46324. I didn't touch `docker_image` because many of its operations cannot work with the image ID / content hash as far as I understand it.

@wrouesnel Any opinions on `docker_image`?

felixfontein commented:
I think this is now resolved by #47393. Thanks for your work!

close_me
